cache/memory/memory.go

```diff
@@ -67,24 +99,18 @@ func newContainerStore(ref info.ContainerReference, maxAge time.Duration) *conta
 }
 
 type InMemoryCache struct {
-	lock              sync.RWMutex
-	containerCacheMap map[string]*containerCache
+	containerCacheMap containerCacheMap
 	maxAge            time.Duration
 	backend           []storage.StorageDriver
 }
 
 func (c *InMemoryCache) AddStats(cInfo *info.ContainerInfo, stats *info.ContainerStats) error {
-	var cstore *containerCache
-	var ok bool
-
-	func() {
-		c.lock.Lock()
-		defer c.lock.Unlock()
-		if cstore, ok = c.containerCacheMap[cInfo.ContainerReference.Name]; !ok {
-			cstore = newContainerStore(cInfo.ContainerReference, c.maxAge)
-			c.containerCacheMap[cInfo.ContainerReference.Name] = cstore
-		}
-	}()
+	name := cInfo.ContainerReference.Name
+	cstore, ok := c.containerCacheMap.Load(name)
+	if !ok {
+		newStore := newContainerStore(cInfo.ContainerReference, c.maxAge)
+		cstore, _ = c.containerCacheMap.LoadOrStore(name, newStore)
```

Comment:
part of me wanted to suggest we only Store here, but I suppose two AddStats can race against each other and we mine as well take the old one once created and not replace...